### PR TITLE
Feat: add STR_TO_TIME parsing and generation for teradata and sqlite

### PR DESCRIPTION
**Pull request by indrajeetadityaroy9**
The PR introduces the functionality of converting strings to date-time or timestamp objects with custom formats in Teradata and SQLite. 

https://dbmstutorials.com/random_teradata/teradata-string-to-timestamp.html
https://www.sqlitetutorial.net/sqlite-date-functions/sqlite-strftime-function/

### CODE REVIEW

File: sqlglot/dialects/teradata.py [hunk 1]
```diff
@@ -7,6 +7,18 @@
 from sqlglot.tokens import TokenType
 
 
+def str_to_date_sql(self: Teradata.Generator, expression: exp.StrToDate) -> str:
+    date_string = self.sql(expression, "this")
+    format_specifier = self.format_time(expression)
+    return f"TO_DATE({date_string}, {format_specifier})"
```
Comment (georgesittas) [inline sqlglot/dialects/teradata.py:13]: Looks like this is not used anywhere, let's get rid of it

File: tests/dialects/test_sqlite.py [hunk 2]
```diff
@@ -140,6 +140,19 @@ def test_sqlite(self):
             read={"snowflake": "LEAST(x, y, z)"},
             write={"snowflake": "LEAST(x, y, z)"},
         )
+        self.validate_all(
+            "SELECT STR_TO_DATE('2022-01-01', '%Y-%m-%d')",
```
Comment (georgesittas) [inline tests/dialects/test_sqlite.py:144]: Take a look at how `validate_all` is [defined](https://github.com/tobymao/sqlglot/blob/3bd811d67e7001046812f2a8590bd07e23b81c88/tests/dialects/test_dialect.py#L27C14-L27C14): the first argument gets parsed according to the dialect corresponding to the surrounding class (here: SQLite). This means that it's generally SQL which is valid in that dialect, i.e. `STRFTIME` instead of `STR_TO_DATE`.

File: sqlglot/dialects/sqlite.py [hunk 3]
```diff
@@ -27,6 +27,16 @@ def _date_add_sql(self: SQLite.Generator, expression: exp.DateAdd) -> str:
...
+def str_to_date_sql(self: SQLite.Generator, expression: exp.StrToDate) -> str:
+    date_string = self.sql(expression, "this")
+    format_specifier = self.format_time(expression)
+    return f"STRFTIME({format_specifier}, {date_string})"
+
+
+def str_to_time_sql(self: SQLite.Generator, expression: exp.StrToTime) -> str:
+    time_string = self.sql(expression, "this")
+    format_specifier = self.format_time(expression)
+    return f"STRFTIME({format_specifier}, {time_string})"
```
Comment (georgesittas) [inline sqlglot/dialects/sqlite.py:40]: These look identical, can we consolidate them into a single function and then reuse it for both `StrToDate` and `StrToTime` to avoid code duplication?

File: tests/dialects/test_sqlite.py [hunk 4]
```diff
@@ -140,6 +140,19 @@ def test_sqlite(self):
...
+                        "mysql": "SELECT STR_TO_DATE('2022-01-01', '%Y-%m-%d')",
+                        "sqlite": "SELECT STRFTIME('%Y-%m-%d', '2022-01-01')"
+                    }
+        )
+        self.validate_all(
+            "SELECT CAST('2022-01-01' AS DATE FORMAT '%Y-%m-%d')",
+            write = {
+                "sqlite": "SELECT STRFTIME('%Y-%m-%d', '2022-01-01')"
+            }
+        )
```
Comment (georgesittas) [inline tests/dialects/test_sqlite.py]: This test is incorrect, perhaps because of a misunderstanding surrounding `validate_all`. The way you've written this is like expecting the two queries to produce the same results because you're both reading and writing SQLite, but if you run them against the engine you'll notice they produce different results.

File: tests/dialects/test_teradata.py [hunk 5]
```diff
@@ -205,3 +205,41 @@ def test_cast(self):
...
+            write={
+                "teradata": "TO_TIMESTAMP('2024-01-01 13:14:15', 'YYYY-MM-DD HH24:MI:SS')",
+                "oracle": "TO_TIMESTAMP('2024-01-01 13:14:15', 'YYYY-MM-DD HH24:MI:SS')",
+                "postgres": "TO_TIMESTAMP('2024-01-01 13:14:15', 'YYYY-MM-DD HH24:MI:SS')",
+            },
+        )
+        self.validate_all(
+            "CAST('2024-01-01' AS DATE FORMAT 'YYYY-MM-DD')",
+            write={"teradata": "CAST('2024-01-01' AS DATE FORMAT 'YYYY-MM-DD')"},
+        )
```
Comment (georgesittas) [inline tests/dialects/test_teradata.py:229]: Nit: we generally try to use `validate_identity(input[, expected])` for tests like this where we're only interested to check the parsing / generation roundtrip for the given dialect, because it's more compact. So this would look like:

```python
        self.validate_identity("CAST('2024-01-01' AS DATE FORMAT 'YYYY-MM-DD')")
```

File: tests/dialects/test_sqlite.py [hunk 6]
```diff
@@ -140,6 +140,19 @@ def test_sqlite(self):
             read={"snowflake": "LEAST(x, y, z)"},
             write={"snowflake": "LEAST(x, y, z)"},
         )
+        self.validate_all(
+            "SELECT STR_TO_DATE('2022-01-01', '%Y-%m-%d')",
+            write = {
+                        "mysql": "SELECT STR_TO_DATE('2022-01-01', '%Y-%m-%d')",
+                        "sqlite": "SELECT STRFTIME('%Y-%m-%d', '2022-01-01')"
+                    }
+        )
```
Comment (georgesittas) [inline tests/dialects/test_sqlite.py]: Run `make style` for commits that are part of a PR so that the CI doesn't crash due to styling / linter issues

File: sqlglot/dialects/teradata.py [hunk 7]
```diff
@@ -202,6 +208,7 @@ class Generator(generator.Generator):
             exp.StrToDate: lambda self, e: f"CAST({self.sql(e, 'this')} AS DATE FORMAT {self.format_time(e)})",
             exp.ToChar: lambda self, e: self.function_fallback_sql(e),
             exp.Use: lambda self, e: f"DATABASE {self.sql(e, 'this')}",
+            exp.StrToTime: lambda self, e: str_to_time_sql(self, e),
```
Comment (georgesittas) [inline sqlglot/dialects/teradata.py:211]:
```suggestion
            exp.StrToTime: str_to_time_sql,
```

File: tests/dialects/test_sqlite.py [hunk 8]
```diff
@@ -140,6 +140,16 @@ def test_sqlite(self):
...
+        self.validate_all(
+            "SELECT STR_TO_DATE('2022-01-01', '%Y-%m-%d')",
+            write={
+                "mysql": "SELECT STR_TO_DATE('2022-01-01', '%Y-%m-%d')",
+                "sqlite": "SELECT STRFTIME('%Y-%m-%d', '2022-01-01')",
+            },
+        )
+        self.validate_identity("STRFTIME('%Y-%m-%d %H:%M:%f', '2024-01-01 13:14:15.123')")
+        self.validate_identity("STRFTIME('%Y-%m-%d %I:%M:%S %p', '2024-01-01 01:14:15 PM')")
+        self.validate_identity("STRFTIME('%d-%b-%Y', '01-Jan-2024')")
```
Comment (georgesittas) [inline tests/dialects/test_sqlite.py:152]: Let's try to increase the coverage here by testing how `STRFTIME` is transpiled to some other important dialects like postgres, mysql, duckdb

Comment (georgesittas) [reply tests/dialects/test_sqlite.py:152]: Also, I'm curious about the last two tests here because SQLite seems to return `NULL` for them or something?

```sql
sqlite> SELECT STRFTIME('%Y-%m-%d %I:%M:%S %p', '2024-01-01 01:14:15 PM');

sqlite> SELECT STRFTIME('%d-%b-%Y', '01-Jan-2024');

```

Are these formats correct? Why are we not getting back a proper value?

File: tests/dialects/test_teradata.py [hunk 9]
```diff
@@ -205,3 +205,28 @@ def test_cast(self):
...
+                "postgres": "TO_TIMESTAMP('2024-01-01 13:14:15', 'YYYY-MM-DD HH24:MI:SS')",
+            },
+        )
+        self.validate_identity("CAST('2024-01-01' AS DATE FORMAT 'YYYY-MM-DD')")
+        self.validate_identity("CAST('2024/01/01' AS DATE FORMAT 'YYYY/MM/DD')")
+        self.validate_identity(
+            "TO_TIMESTAMP('2024-01-01 13:14:15.123', 'YYYY-MM-DDBHH:MI:SS.S(3)')"
+        )
+        self.validate_identity("TO_TIMESTAMP('2024-01-01 01:14:15 PM', 'YYYY-MM-DDbHH:MI:SSBT')")
+        self.validate_identity("CAST('01-Jan-2024' AS DATE FORMAT 'DD-MMM-YYYY')")
```
Comment (georgesittas) [inline tests/dialects/test_teradata.py:232]: Let's also try to increase the coverage here - the identity `CAST(... AS DATE FORMAT ...)` tests don't seem to be testing any new functionality here, we already have a few tests for this here: https://github.com/tobymao/sqlglot/blob/main/tests/dialects/test_teradata.py#L183

**Review by georgesittas — COMMENTED** [review]
Left a few more comments.
